col/src/main/java/vct/col/ast/util/NameScanner.scala

```diff
@@ -0,0 +1,210 @@
+  def freeVars[R <: ASTNode](nodes: util.List[R]): util.Map[String, Type] = {
+    val scanner = new NameScanner
+    for (n <- nodes.asScala) {
+      n.accept(scanner)
+    }
+    scanner.freeNamesJava
+  }
+
+  @varargs
+  def freeVars(theNodes: ASTNode*): util.Map[String, Type] = freeVars(theNodes.asJava)
+}
+
+class NameScanner extends RecursiveVisitor[AnyRef](null, null) {
+
+  case class Entry(name: String, typ: Type, var writtenTo: Boolean)
+
+  val frameStack: mutable.Stack[mutable.Set[DeclarationStatement]] = mutable.Stack()
+
+  // I suspect there is some external code that depends on insertion order... Therefore use LinkedHashMap
+  val freeNames: mutable.Map[String, Entry] = mutable.LinkedHashMap()
+
+  // Ensure there is at least one frame at the start
+  push()
+
+  private def hasDecl(name: String): Boolean = frameStack.exists(frame => frame.exists(p => p.name == name))
+
+  private def put(name: String, typ: Type): Unit = {
+    if (hasDecl(name)) {
+      Abort("Cannot put a free var when it has a decl")
+    }
+
+    Objects.requireNonNull(typ)
+
+    freeNames.get(name) match {
+      case Some(_) =>
+        // Disallow putting the same free var again, to prevent the writtenTo flag to be accidentally reset
+        Abort("Name %s is already marked as free", name);
+      case None =>
+        freeNames.put(name, Entry(name, typ, writtenTo = false));
+    }
+  }
+
+  private def setWrite(name: String): Unit =
+    freeNames.get(name) match {
+      case Some(value) => value.writtenTo = true
+      case None => Abort("Cannot set write for non-existent variable %s", name)
+    }
+
+  private def push(): Unit = frameStack.push(mutable.Set())
+
+  private def pop(): Unit = frameStack.pop()
+
+  /**
+    * Returns free names as a java map with insertion order retained
+    */
+  def freeNamesJava: util.Map[String, Type] = {
+    val map = new util.LinkedHashMap[String, Type]()
+    freeNames.foreach(mapEntry => map.put(mapEntry._1, mapEntry._2.typ))
+    map
+  }
+
+  /**
+    * All variable names that are read from or written to
+    * @return
+    */
+  def accesses: Set[String] = freeNames.keySet.toSet
+
+  def writes: Set[String] = freeNames.filter(p => p._2.writtenTo).keySet.toSet
+
+  def reads: Set[String] = freeNames.filter(p => !p._2.writtenTo).keySet.toSet
+
+  override def visit(e: NameExpression): Unit = e.getKind match {
+    case Reserved =>
+    case Label =>
+    case Field | Local | Argument =>
+      val name = e.getName
+      val t = e.getType
+
+      freeNames.get(name) match {
+        case Some(value) =>
+          // If the name is already a free variable, the types must match
+          if (t != null && !(value.typ == t)) {
+            Fail("type mismatch %s != %s", t, value.typ);
+          }
+        case None if !hasDecl(name) =>
+          // If it is new and there is no decl, declare a new free variable
+          put(name, t)
+        case _ =>
+          // Otherwise ignore it
```

Comment:
I don't understand what is happening here, but if this was the old situation let's just leave it.

Reply:
I've rewritten the match as there was also an edge case missing and added a comment, hopefully it is clearer now